---
title: "<center>Biostatistics - BIOL 260"
author: "<center>*Lab 3: Visualizing data*"
output:
  html_document: default
  pdf_document: default
---

<style type="text/css">
.main-container {
  max-width: 800px;
}
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
/* Whole document: */
body{
  font-family: helvetica;
  font-size: 12pt;
}
code {white-space:pre !important; overflow-x:scroll !important;font-size:12pt; font-family:courier;}
</style>


<br><br>
<style>
body {
text-align: justify}
div.blue { background-color:#e6f0ff; border-radius: 5px; padding: 20px;}
</style>
<div class = "blue">

**Today's investigation.** Displaying data using appropriate tools (graphs) is one of the most important jobs we have as scientists. Before any data analysis, we need to understand the patterns in our data. Moreover, good data visualization brings us the opportunity to show the newly discovered trends and relationships between variables. That is, through good data visualization we may reach different audiences and send a clear message home. Today, we will learn how to generate graphs parting from the type of data we have and the research question we want to answer using two different datasets of marine invertebrates from CSULB Marine Ecology Lab and of microbial genomics from CSULB Comparative Microbial Genomics Lab.
</div>

***

<span style='color:#66023C; font-size:150%; font-weight:600'> Introduction </span>
<br>

In this lab we will generate multiple graphs to identify good data visualization practices, which include data displays that are clear, easy-to-interpret, and unbiased. When we carry out an experiment, a field study or a computer-based analysis, we are evaluating variables. These variables can take different forms; they can be *categorical* (e.g., males vs females; climatic seasons) or *numerical* (e.g., weight; temperature), and our main interest may be to show a *range of values for one variable* or an *association between two variables*. So, how do we choose between the many data visualization tools available in our toolbox (e.g., bar plot, box plots, line plots, scatter plots, etc.)? 

Today, we will explore different biological datasets in order to generalize on good data visualization practices. Specifically, we will practice how to identify variable types and explore effective ways to display such variables using anatomical data from the crab *Uca pugnax* collected by the CSULB Marine Ecology Lab lead by Dr. Bengt Allen and microbial genomic data extracted from the public database PATRICbrc and used by CSULB Comparative Microbial Genomics Lab lead by Dr. Renaud Berlemont (Fig 1). The R programming language offers multiple ways to carry out high-quality graphs, however, it is up to us to decide which one to use for our data. So, let's first explore the different datasets and decide the most appropriate graphs to display it, while learning how to code them in R. Then, in the Take-home exercise we will generate a research question and carry out data exploration to answer it with graphs.

```{r echo=FALSE,fig.align="center",fig.asp=.5,fig.width = 6,fig.cap="*Uca* species. Image: Dominique DeLambert"}
library(png)
library(grid)
img <- readPNG("uca_pugilator_Dominique_DeLambert.png")
grid.raster(img)
```
<br>

**Upon completion of this lab, you should be able to:**

  + Differentiate between variable types;
  + Identify appropriate forms of data visualization given a set of variables;
  + Read and interpret graphs.

<br>

**External resources**

  + [Pathosystems Resource Integration Center (PATRIC)](https://patricbrc.org)
  + [Fundamentals of Data Visualization](https://serialmentor.com/dataviz/)

***

<span style='color:#66023C; font-size:150%; font-weight:600'> Worked example </span> 
<br>

To get started, let’s remind ourselves the different variable types:

  + categorical - describes membership in a category or group. These variables do not correspond to a numerical magnitude. 
  + numerical - describes a quantitative measure. These variables are numbers and thus have magnitude in a numerical scale. Numerical variables can also be continuous or discrete. A *continuous numerical variable* can take any value within a range, while a *discrete numerical variable* can take integer values only.
<br><br>

When we are exploring associations between variables, they are also categorized into explanatory or response variables:

  + explanatory variable - the variable that predicts or affects the other. This variable is also known as the *independent* variable.
  + response variable - the variable predicted or affected by the explanatory variable. This variable is also known as the *dependent* variable.
<br><br>

  
Our choice for data visualization depends on both the types of variables and our research question. However, there are many other aesthetics that are fully dependent on personal preference. Thus, we often have more than one choice. Say that we are exploring 
data from the National Climatic Data Center [NCDC](https://www.ncdc.noaa.gov/data-access/land-based-station-data/land-based-datasets/climate-normals), which provides the daily temperature normals (average daily temperatures over a 30-year window) for different US sites, because we are interested in knowing what periods of the year are characterized by less variability in temperature across four US sites (Chicago, Death Valley, Houston, and San Diego). In this case, temperature is a *continuous numerical* variable and days is a *continuous discrete* variable (takes integer values of 1 to 366, exclusively). Because we are interested in how temperature changes across days, we say that day is the *explanatory* variable affecting temperature, and temperature is the *response* variable being affected by day of the year. 

Let's explore three different visualizations for the same temperature data and how they are useful for different messages.  
<br>

### Visualization 1: How does site-specific temperature change across time?
A common way to display trends in time, such as temporal changes in temperature, is through *line graphs*. Usually, we display one data point of the *y*-axis (response=temperature) per each data point in time (explanatory=days), which is displayed in the *x*-axis. In this way, adjacent points along the *x*-axis are connected by a line, displaying the temporal trend (here labeled by month). 
<br><br>

```{r,echo=FALSE, fig.align='center',message=FALSE,warning=FALSE,fig.asp=.4,fig.width=7,fig.cap="Figure 1. Daily temperature normals for four locations in the U.S. Data source: NOAA."}
library(ggplot2)
library(tidyverse)
library(lubridate)
tidy_normals <- read.csv("tidy_normals.csv",header=TRUE)

# converting F to C
convert_fahr_to_kelvin <- function(temp) {
  kelvin <- ((temp - 32) * (5 / 9)) + 273.15
  return(kelvin)
}

convert_kelvin_to_celsius <- function(temp) {
  celsius <- temp - 273.15
  return(celsius)
}

fahr_to_celsius <- function(temp) {
  temp_k <- convert_fahr_to_kelvin(temp)
  result <- convert_kelvin_to_celsius(temp_k)
  return(result)
}

tidy_normals <- read_csv("tidy_normals.csv", col_types = cols(
  station_id = col_character(),
  month = col_double(),
  day = col_double(),
  temperature = col_double(),
  flag = col_character(),
  date = col_date(),
  location = col_factor(),
  month_name = col_factor()
  )) %>%
  mutate(location = fct_relevel(location,
                                 "Death Valley",
                                 "Houston",
                                 "San Diego",
                                 "Chicago"
                                 ))

temps_cartesian <- ggplot(tidy_normals, aes(x = date, y = fahr_to_celsius(temperature), color = location)) +
  geom_line(size = 1.5) +
  scale_x_date(name = "Month", limits = c(ymd("0000-01-01"), ymd("0001-01-04")),
               breaks = c(ymd("0000-01-01"), ymd("0000-04-01"), ymd("0000-07-01"),
                          ymd("0000-10-01"), ymd("0001-01-01")),
               labels = c("Jan", "Apr", "Jul", "Oct", "Jan"), expand = c(1/366, 5)) +
  scale_y_continuous(limits = c(-10, 40),
                     breaks = seq(-10, 400, by = 20),
                     name = "Temperature (°C)") +
  theme_classic(15) +
  theme(panel.border = element_blank(),
        panel.grid.minor = element_blank(),
        legend.title = element_blank(), 
        legend.position = "top",
        legend.text=element_text(size=14),
        )

temps_cartesian
```
<br>

**Interpretation.** From this line graph, we can observe general patterns in temperature across time; cooler temperatures during winter months and warmer temperatures during summer months. We can also observe site-specific patterns because the data also includes a third categorical variable; *location*. According to the displayed data, Chicago has more variability in daily temperature normals, while San Diego has less variability. This simple data visualization exercise opens up more interesting questions concerning latitudinal patterns in temperature and the drivers of variability in such latitudinal gradient.

<br>

### Visualization 2: What month exhibits the highest mean temperature variation across sites?
Another visual technique that we can use to show the magnitude and pattern of a phenomenon are *heat maps*. In this case, we display a variation in color that corresponds to the magnitude of the response variable. This serves as a visual cue to the reader about how the variable changes and/or is clustered over time.  
<br><br>

```{r,echo=FALSE,out.width="100%",fig.align='center',fig.asp=.30,fig.cap="Figure 2. Monthly normal mean temperatures for four locations in the U.S. Data source: NOAA"}

mean_temps <- tidy_normals %>%
  group_by(location, month_name) %>%
  summarize(mean = mean(fahr_to_celsius(temperature)))

temps_heatmap <- ggplot(mean_temps, aes(x = month_name, y = location, fill = mean)) + 
  geom_tile(width = 0.95, height = 0.95) +
  coord_equal() +
  scale_fill_viridis_c(option = "B", begin = 0.15, end = 1,
                       name = "Temperature (°C)") + 
  xlab("Month") +
  ylab(NULL) +
  theme_bw(15) +
  theme(panel.border = element_blank(),
        panel.grid.major = element_blank(),
        legend.title = element_text(size=10),
        legend.text=element_text(size=14),
        axis.line = element_blank(),
        axis.ticks = element_blank()
        )

temps_heatmap
```
<br>

**Interpretation.** From this heat map, we can infer that spring and fall months exhibit the highest variation in mean temperature normals (they depict a larger range of colors representing mean monthly temperature). Although we could have inferred this same message from Fig. 1, the heat map makes it easier to our eyes. Another message we can get from this graph is the repetitive pattern that mean monthly temperature has across these sites. That is, the figure appears to have a mirror image in its center reflecting similar sides.   

<br>

### Visualization 3: How periodic is daily temperature across sites?
A last visualization that is useful for periodic data, such as daily temperature across years, are *polar coordinates*. Here, data values at one end of the scale can be logically joined to data values at the other end. Let's consider our explanatory variable; days. December 31st is the last day of the year, but it is also one day before the first day of the year. If we want to show how some quantity varies over the year, we can use polar coordinates with the angle coordinate specifying each day. Here, the radial distance from the center point indicates the daily temperature and the days of the year are arranged counter-clockwise starting with Jan. By plotting the temperature normals in a polar coordinate system, we emphasize the cyclical property they have. 
<br>


```{r,echo=FALSE,fig.align='center',fig.cap="Figure 3. Daily temperature normals for four selected locations in the U.S. Data source: NOAA"}
temps_polar <- ggplot(tidy_normals, aes(x = date, y = fahr_to_celsius(temperature), color = location)) +
  geom_line(size = 2) +
  scale_x_date(name="Date") + 
  scale_y_continuous(limits = c(-10, 40),
                     breaks = seq(-10, 40, by = 20),
                     name = "Temperature (°C)") +
  theme_bw(21) +
  coord_polar(theta = "x", start = pi, direction = -1) +
  theme(panel.border = element_blank(),
        panel.grid.minor = element_blank(),
        legend.title = element_blank(),
        legend.position = "top",
        legend.text=element_text(size=14),
        axis.line = element_blank(),
        axis.ticks = element_blank()
        )
temps_polar
```
<br>
 
**Interpretation.** From this polar coordinate, we can address directly the repetitiveness and thus periodicity that temperature has through time. If this was not the case, we would not observe circles in this plot. Similar to Fig. 1 and Fig. 2, we can also see that Death Valley presents the warmest temperatures during summer, while Chicago exhibits the coldest temperatures during winter, and that Death Valley, Houston and San Diego present similar temperatures during winter.

***

<br>
<span style='color:#66023C; font-size:150%; font-weight:600'> Materials and Methods </span> 
<br>

  + R and RStudio
  + R packages *ggplot2* and *tidyverse*
  + uca dataset
  + micro dataset 
<br><br>

Today's methods are organized into five main exercises using the relatively small *Uca* crab dataset followed by a challenge exercise using the larger microbial genomics dataset. 

<br>
Import both datasets to RStudio and check them out. **Don't forget to use the metadata file for reference.**
```{r,eval=FALSE}
# importing uca crab data
uca <- read.csv("uca.csv",header=TRUE)
View(uca)

# importing microbial genomics data
micro <- read.table("micro.txt")
View(micro)
```
<br>

**Questions**

  a. *How many variables and observations "uca" has?* **Hint**: review past R scripts!
  b. *How many variables and observations "micro" has?*
<br><br>

<style>
body {
text-align: justify}
div.creme { background-color:#fff5e6; border-radius: 5px; padding: 20px;}
</style>
<div class = "creme">

<span style='color:blue; font-size:120%; font-weight:600'> **Info-Box!** </span> To plot the data, we will use R package *ggplot2*. To have a preview of plots, click on “Plots” in the bottom right window of RStudio. In *ggplot2*, we use the function **ggplot()** to generate the graphs. The basic template for ggplots is 

<center>
<span style='color:blue; font-size:100%; font-weight:450'> ggplot(data = DATA, mapping = aes(MAPPINGS)) +  GEOM_FUNCTION()</span>, </center>
<br>
where DATA is our data, MAPPINGS are the *x*-axis and *y*-axis variables, and GEOM_FUNCTION is the graph type. *ggplot2* offers many different geoms, including:

  + **geom_bar()**, for bar graphs;
  + **geom_histogram()**, for histograms;
  + **geom_point()**, for scatter plots;
  + **geom_boxplot()**, for boxplots;
  + **geom_line()**, for trend lines and time series.
</div>
<br>


**1. Displaying data for one variable**

To examine data for a single variable, we use **frequency distributions**, which are the number of occurrences of each value in the data.

<br>

**A. When displaying categorical data, use a bar graph**. 


A bar graph uses the *height* of the bars to display the frequency distribution of categorical variables. Say we are interested in exploring the claw size distribution of the *Uca* crab. Are there size classes with more representation in the population? To answer this question, we can generate a frequency distribution for the different claw size classes (categorical variables; i.e., small, medium, large) using a bar graph.

```{r,eval=FALSE}
# revising the data frame by checking the first rows of the data frame using head()
head(uca)

# loading package ggplot2
library(ggplot2)

# frequency of claw size classes
p1 <- ggplot(data=uca,aes(x=claw_size)) +
  geom_bar() 
p1
```
<br>

A good practice is to order the bars by magnitude as it helps the reader to find important patterns in the frequency data. Let's reorder the frequency bars from highest magnitude to lowest. For this, we first need to estimate a frequency table for claw sizes using **table()**.
```{r,eval=FALSE}
# frequency table of claw size classes
freq_t <- table(uca$claw_size)
freq_t

# converting freq_t into a data frame to plot it
freq_t <- as.data.frame(freq_t)
freq_t

# reordering the bars per magnitude in ggplot using the new data frame created and the function reorder() 
p1 <- ggplot(data=freq_t,aes(x=reorder(Var1,-Freq),y=Freq)) +
   geom_bar(stat="identity")
p1
```
<br>

Finally, let's add some important aesthetics to *p1*.
```{r,eval=FALSE}
# adding y- and x-axis labels using ylab() and xlab()
p1 <- ggplot(data=freq_t,aes(x=reorder(Var1,-Freq),y=Freq)) +
  geom_bar(stat="identity") +
  ylab("Frequency") +
  xlab("Claw size")
p1

# deleting the background color using theme_classic()
p1 <- ggplot(data=freq_t,aes(x=reorder(Var1,-Freq),y=Freq)) +
  geom_bar(stat="identity") +
  ylab("Frequency") +
  xlab("Claw size") +
  theme_classic()
p1

# increasing the font size to "20"
p1 <- ggplot(data=freq_t,aes(x=reorder(Var1,-Freq),y=Freq)) +
  geom_bar(stat="identity") +
  ylab("Frequency") +
  xlab("Claw size") +
  theme_classic(20)
p1
```
<br>

**Questions**

  a. *What type of variable is claw_size?*
  b. *Are claw size classes represented equally in the population?* 
  c. *Which size class is least represented in the population?*
<br><br>



<span style='color:red; font-size:150%; font-weight:600'> Challenge 1. </span> Say we are interested in exploring biases in available microbial sequenced genomes. Are there particular microbes that have been sequenced by scientists more than others? To answer this question we can generate a frequency distribution for different microbial classifications (categorical variables; e.g., kingdom, phylum) using also bar graph. **Reproduce the bar graph below showing the number of sequenced genomes per kingdom**  (categorical variables; i.e., Archaea, Bacteria, Viruses) using *p1* as example. 

```{r echo=FALSE,fig.align="center",fig.asp=.4,fig.width = 9,fig.cap="Figure 2. Microbial sequenced genomes per kingdom. Data source: PATRICbrc dataset."}
library(png)
library(grid)
img <- readPNG("image1.png")
grid.raster(img)
```

<br>


**B. When displaying numerical data, use a histogram**. 


In contrast to bar graphs, histogram use the *area* of the bars to display the frequency distribution of numerical variables. In this way, histograms split the data values into intervals or **bins**, reveling the shape of frequency distributions. 

Let's explore the claw length distribution of the *Uca* crab.
```{r,eval=FALSE}
# histogram for claw length distribution
p2 <- ggplot(uca,aes(x=claw_length)) +
  geom_histogram() 
p2

# changing the bin width
p2 <- ggplot(uca,aes(x=claw_length)) +
  geom_histogram(binwidth = 2)
p2

# adding aesthetics
p2 <- ggplot(uca,aes(x=claw_length)) +
  geom_histogram(binwidth = 2) +
  ylab("Frequency") +
  xlab("Claw length (mm)") +
  theme_classic(20)
p2
```
<br>

**Questions**

  a. *What type of variable is claw_length?*
  b. *What happens as you increase the bin width?* 
  c. *What shape does the frequency distribution of claw length has?*
<br><br>


<br>
<span style='color:red; font-size:150%; font-weight:600'> Challenge 2. </span> Now, let's say we are interested in exploring the genome length distribution of the available microbial sequenced genomes. **Reproduce the histogram below showing the microbial genome length distribution across kingdoms** using *p2* as example. 

<br>

```{r echo=FALSE,fig.align="center",fig.asp=.4,fig.width = 9,fig.cap="Figure 3. Microbial sequenced genomes per kingdom. Data source: PATRICbrc dataset."}
library(png)
library(grid)
img <- readPNG("image2.png")
grid.raster(img)
```
<br>

**2. Displaying associations between two variables**

Here, we are interested in graphing two different variables to show an association or correlation, or a difference between groups. 

<br>

**A. When displaying categorical data, use a mosaic plot**. 

A mosaic plot is similar to a bar graph but bars within groups are stacked on top of one another.


Let's explore associations between claw size and crab mass using a mosaic plot.
```{r,eval=FALSE}
# mosaic plot for claw size and crab body mas
p3 <- ggplot(data=uca,aes(x=claw_size,fill=mass_class)) +
  geom_bar() +
  ylab("Frequency") +
  xlab("Claw size") +
  theme_classic(20)
p3
```
<br>

**Questions**

  a. *What type of variable are claw_size and mass_class?*
  b. *What associations between claw size and crab body mass can you infer from the graph?* 
  c. *What distribution of mass classes would you expect in a population of crabs with large claw size?*
<br><br>


<span style='color:red; font-size:150%; font-weight:600'> Challenge 3. </span> Say we are interested in exploring associations between available sequenced genomes per kingdom and phylum. **Reproduce the mosaic plot below showing the number of microbial sequenced genome per kingdom and phylum** using *p3* as example.

<br>

```{r echo=FALSE,fig.align="center",fig.asp=.4,fig.width = 9,fig.cap="Figure 4. Microbial sequenced genomes per kingdom and phylum. Data source: PATRICbrc dataset."}
library(png)
library(grid)
img <- readPNG("image3.png")
grid.raster(img)
```

<br>


**B. When displaying numerical data, use scatter plots**. 

A scatter plot is used whenever we want to explore the association between two numerical variables. Here, the explanatory variable is positioned in the *x*-axis, while the response variable is positioned in the *y*-axis.

Let's explore the association between *Uca*'s claw length and claw mass using **geom_point()**.

```{r,eval=FALSE}
# scatter plot for claw length and claw mass 
p4 <- ggplot(uca,aes(x=claw_mass,y=claw_length)) +
  geom_point()
p4

# differentiating by mass class adding the aesthetic of color=mass_class
p4 <- ggplot(uca,aes(x=claw_mass,y=claw_length,color=mass_class)) +
  geom_point()
p4

# adding all aesthetics
p4 <- ggplot(uca,aes(x=claw_mass,y=claw_length,color=mass_class)) +
  geom_point() +
  ylab("Claw length (mm)") +
  xlab("Claw mass (g)") +
  theme_classic(20)
p4
```
<br>

**Questions**

  a. *What type of variable are claw_mass, claw_length, and mass_class?*
  b. *What is the response variable and the explanatory variable?*
  c. *What associations can you infer from the graph?* 
<br><br>


<span style='color:red; font-size:150%; font-weight:600'> Challenge 4. </span> Say we are interested in exploring whether longer microbial genomes are associated to larger number of genes in the genome. **Reproduce the scatter plot below showing the association between the number of genes and the genom length for microbes across phylum** using *p3* as example. **Hint:** You may plot the data without a legend by adding **theme(legend.position = "none")** to the plot codes.

<br>


```{r,eval=FALSE, echo=FALSE}
ggplot(micro,aes(x=genome_length,y=patric_cds,color=phylum)) +
  geom_point() +
  ylab("Coding DNA sequences or genes") +
  xlab("Genome length (nucleotides)") + 
  theme_classic(20) +
  theme(legend.position = "none")
```

```{r echo=FALSE,fig.align="center",fig.asp=.4,fig.width = 9,fig.cap="Figure 5. Number of genes as a function of genome length. Data source: PATRICbrc dataset."}
library(png)
library(grid)
img <- readPNG("image4.png")
grid.raster(img)
```

<br>

**C. When displaying numerical and categorical data, use box plots**. 

A box plot uses lines and rectangular boxes to display a summary of the frequency distribution of the variable of interest. Specifically, box plots show the median, quartiles, range, and extreme measurements of the data. Let's first review the definitions of each component in the summary:

  + *median* - the middle measurement of the observations. That is, half of the observations are located below the median and the other half are located above the median.   
  + *quartile* - each of four equal groups into which the data can be divided according to the distribution of values. That is, one-fourth or 25% of the data lies below the *first quartile* and one-fourth or 25% of the data lies above the *third quartile*. 
  + *range* - the difference between the maximum and minimum data value. 
  + *extreme measure* - a data point that differs significantly from the rest.
  
```{r,echo=FALSE,fig.align="center",fig.height=5.5,fig.width=4.5, fig.cap="Figure 6. Anatomy of a box plot."}
library(ggplot2)
df <- as.data.frame(c(7.3,7.8,10.2,9.8,8.5,11.2,8.8,7.1,9,13.6))
df$cat <- "Group"
names(df) <- c("Value","Categories")

ggplot(df,aes(Categories,Value)) +
  geom_boxplot(fill="lightgreen") +
  geom_segment(aes(x = 1, y = 10.1, xend = 1.5, yend = 10.1),color="blue") +
  geom_segment(aes(x = 1, y = 7.975, xend = 1.5, yend = 7.975),color="blue") +
    geom_segment(aes(x = 0.65, y = 8.9, xend = 1.35, yend = 8.9),color="orange") +
  annotate("text", x=1.55, y=8, label= "Q1",size=6) +
  annotate("text", x=1.55, y=10.1, label= "Q3",size=6) +
  annotate("text", x=1, y=9.1, label= "median",size=6) +
  annotate("text", x=1.13, y=7.45, label= "whisker",size=5) +
  annotate("text", x=1.13, y=10.55, label= "whisker",size=5) +
  annotate("text", x=.94, y=11.1, label= "max",size=4) +
  annotate("text", x=.94, y=7.15, label= "min",size=4) +
  annotate("text", x=1.3, y=13.62, label= "extreme measure",size=5) +
  ylab("") +
  xlab("") +
  theme_classic(20) +
  theme(
    panel.border = element_blank(), 
    axis.line.x = element_blank(), 
    axis.line.y = element_blank(), 
    axis.text.y = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.y = element_blank(),
    axis.ticks.x = element_blank())
```


Let's explore the summaries of the frequency distributions of *Uca* crabs body mass across claw size classes.
```{r,eval=FALSE}
# box plot for body mass across claw size using geom_boxplot
p5 <- ggplot(uca,aes(x=claw_size,y=crab_mass)) +
   geom_boxplot()
p5

# adding aesthetics
p5 <- ggplot(uca,aes(x=claw_size,y=crab_mass,fill=claw_size)) +
   geom_boxplot() +
  ylab("Body mass (g)") +
  xlab("Claw size class") +
  theme_classic(20)
p5
```

<br>

**Questions**

  a. *What shape does the frequency distribution of body mass has across claw sizes?*
 
<br><br>

<span style='color:red; font-size:150%; font-weight:600'> Challenge 5. </span> Say we are interested in exploring the frequency distribution of gene density (# genes per 1000 nucleotides) across classes in the phylum Actinobacteria.  **Reproduce the box plot below showing the frequency distribution of gene density across Actinobacteria** using *p4* as example. **Hints:** Review Lab 1's script on how to filter data (you will need to create a new data object with Actinobacteria data only). You may rotate the *x*-axis labels by 90 degrees by adding **theme(axis.text.x = element_text(angle = 90)** to the plot codes.

<br>

```{r,echo=FALSE,fig.align="center",fig.cap="Figure 7. Gene density in Actinobacteria.",fig.height=5,fig.width=7}
library(tidyverse)
micro <- read.table("micro.txt")
acti <- filter(micro,
               phylum=="Actinobacteria")

ggplot(acti,aes(x=class,y=density*1000,fill=class)) +
  geom_boxplot() +
  theme_classic() +
  ylab("Gene density (genes per 1000 bases)") +
  xlab("Class") +
  theme_classic(15) +
  theme(legend.position = "none") +
  theme(axis.text.x = element_text(angle = 90))
```

<br>


<style>
body {
text-align: justify}
div.creme { background-color:#fff5e6; border-radius: 5px; padding: 20px;}
</style>
<div class = "creme">

<span style='color:blue; font-size:120%; font-weight:600'> **Info-Box!** </span>  Graphs in any report or manuscript need to fulfill the following general guidelines:

1) **Name and description** - All graphs must be accompanied by a figure name and description (a short title and description below the figure). Note that all the figures in this module have their own name and description. In this way, graphs are stand alone figures in the report. The name of the figure is useful when referencing it in the main text of a report. 

2) **Size and resolution** - All graphs should be easy to read with an appropriate resolution. RStudio provides an easy way to export images using different formats, including .pdf files. This feature is located in the **Plots** window (bottom right window of RStudio). From the Export feature in Plots, you may manipulate the size of the image, the name of the image file and the directory for the saved image file.
</div>
<br>


<center>
**Great Work!**
</center>
<br>

*** 



<span style='color:#66023C; font-size:150%; font-weight:600'> Take-home exercise </span> 
<br> 

<span style='color:#66023C; font-size:100%; font-weight:600'> **Stop, Think, Do:** </span> Now, it is your turn to design your own visualizations. **Stop** and review the different types of variables you have. **Think** about the best practices for good data displays given a particular set of variables. **Do** data explorations by graphing the data that best answers your research question!
<br> 

**Task.** Generate a research question for **each** dataset, one for the marine invertebrate data and one for the microbial genomics data, in [OneDrive](). Once you decide your question, create an annotated R script that includes the following steps; (1) data import, (2) package loading, (3) codes for both plots answering each of your questions. Accompany your R script with a concluding paragraph summarizing your results. Your assignment should include two files; (1) a .doc file with the figures and paragraph, (2) an .R file containing your annotated R script. **Hint:** use your past R scripts and Info-Box! for guidance.   
